### ISSUE TITLE
Gajim sends not OTR messages to my Pidgin

### ISSUE THREAD

**Issue by norpol** [python-otr/gajim-otr#13]
Hi, a friend of mine and me are freaking out. We've got an OTR session running, I can send him encrypted messages but his Gajim client sends me (most of the time) non OTR messages. It work's from time to time, and at least I have no clue why this happens. It's really annoying, since Gajim don't even report on his site that the message wasn't sent encrypted. If this is a libpurple bug you could still implement a feature which notifies the sender that his/her message wasn't encrypted.
<3 - norpol

**Comment by Gjum** [comment]
Has your friend installed and enabled the OTR plugin? Also which version is he running? Because it works for me.

**Comment by wilhelmy** [comment]
Yes. (I'm the friend in question) -  I tried the latest version of gajim-OTR that's on ftp.gajim.org and when that failed I decided to check out the git HEAD. Both have the problem and this happens irregularly and only with particular contacts - I have to check the XML console whether or not I'm sending encrypted messages. Going offline and back online has an effect on it, but I haven't yet been able to single out the failing condition, which is why I didn't report this bug myself yet (I sent a mail to @afflux but he told me he's no longer responsible).

I think you should really add an assertion that only encrypted messages are ever sent as part of a running OTR session, and discard the message with a warning if it would have been sent unencrypted, as well as fixing the actual bug, since this is security software and people might rely on their messages being encrypted.

**Comment by wilhelmy** [comment]
One interesting aspect is that the messages sent as part of such an erratic session contain the OTR-autonegotiation whitespace trailer, which indicates that they indeed pass through gajim-OTR which seems to think there is no session currently active. I think this makes it reasonable to assume that it's not a gajim-core bug (something like a callback not being invoked properly) but a bug in gajim-OTR.

I'll look into it, if nobody beats me to it. I won't promise I'll be able to fix it.

**Comment by afflux** [comment]
This may be an issue related to the way gajim handles sessions / resources. I have a few pointers for you to get the debugging going.

In `src/potr/otrmodule.py` check out `handle_outgoing_msg()`. You should be able to get the output of the potrrootlog with `--loglevel .potr=DEBUG`. If you get varying `fjid`s you will end up using different OTR contexts, which will have different states. You may want to watch the `event.fjid` in `handle_incoming_msg()` as well. Otherwise you can just dump the states of all the contexts:

```python
for uid, ctx in self.us[event.account].ctxs.items():
    potrrootlog.debug('%s:%d', uid, ctx.state)

# STATE_PLAINTEXT = 0
# STATE_ENCRYPTED = 1
# STATE_FINISHED = 2
```

**Comment by ScreenDriver** [comment]
I also report this at the [Gajim issue tracker](https://trac-plugins.gajim.org/ticket/91#comment:1) almost a year ago.

**Comment by norpol** [comment]
@ScreenDriver We found a way which increases the chance of an OTR session. I (with Pidgin) need to close Pidgin, open it again and wait until the Gajim user sends and initiates an OTR session. After this chances are pretty high I get encrypted messages.

**Comment by ScreenDriver** [comment]
@norpol thank you for that information but that's to complicated for the usual users out there. My workaround at the moment is to disable OTR completely :unamused:

**Comment by norpol** [comment]
If I disable auto initiate otr session at my pidgin otr works. :D

**Comment by GreenLunar** [comment]
Related comment at [#91 (OTR plugin does not always encrypt) – Gajim Plugins](https://trac-plugins.gajim.org/ticket/91).

**Comment by pannal** [comment]
This is also the case for me.

When I start a chat with a friend of mine using Psi+ 0.16.261 - my settings enforce OTR for that user in any aspect available - it goes like follows:

Me: "Hello"
Gajim: ‎[16:47:38] ‎[OTR] authenticated secured OTR conversation with asdf@domain/Psi+ started
Me: "Repost the last line you've seen, please"
Him: "Me: Repost the last line you've seen, please"

Offtopic: I'd expect a repost of my original unencrypted message

Sometimes he sees me actually sending an unencrypted message - which was completely disabled on my end's OTR settings for that contact.

To actually make OTR usable in Gajim (especially for the forced mode ("Encryption required"), the OTR plugin should *never* send unencrypted messages *and* re-send them after the encrypted session has started in that case.

**Comment by norpol** [comment]
Yes, we all know. This project is looking for a maintainer, and someone should also file a bug report in their distribution, so people might be notified when installing/using a broken package. See also the README file of this project.

**Comment by norpol** [comment]
I've just filed a report in [Debian #810906](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=810906)

**Comment by kitterma** [comment]
FYI, I've just closed that bug since Debian doesn't ship gotr.

**Comment by norpol** [comment]
@kitterma, thank you for your checking it out. (Sorry) I was wrong, I thought it would be shipped this way, but I've now reinstalled Gajim again and it doesn't contain the plugin by default. (`python-potr` is in the same unmaintained status btw., so it might also not fit into the repository as well)

People can get the plugin via. Gajim's "Plugin Installer".
When selecting "Off-The-Record Encryption" on the list, you are told to read the wiki first:
> Provide OTR encryption. Read https://github.com/python-otr/gajim-otr/wiki before use. 

If you are curios enough and you go to the URL first, you'll read in bold letters:
> **What's the status of python-otr and gajim-otr on GitHub or the gajim-plugin repository?**
> **It's dead.** I don't work on it any longer and have no time or contacts to test any pull request you send to me. If you want to take it over, feel free to let me know or just go ahead and fork. However, I do not recommend using this software, as it's completely unaudited and written by non-cryptographers. I would not be surprised if it provides absolutely no security in any sense.

**In this state, this software could harm people who rely on OTR!**

We need to figure out in which way we should file a bug report for Plugin Installer to not contain the OTR plugin or to find a new maintainer.

PS: https://aur.archlinux.org/packages/gajim-plugin-otr/ should also be issued.

**Comment by lierdakil** [comment]
I'd wager $10 gajim-otr loses OTR session when other end changes resource/priority. Seriously, I'd be happy to send those $10 to anyone who will prove me wrong on this conjecture.

This also opens an interesting attack possibility. Consider this scenario:
1) Alice and Bob established OTR session with alice@jabber.org/Gajim and bob@jabber.org/Gajim, and Alice writes messages to bob@jabber.org (without specifying resource)
2) Eve now connects as bob@jabber.org/Eve with higest resource priority she can think of and with no OTR support (we assume Eve has either Bob's credentials, or control over jabber.org)
3) Best case, Alice sees 'bob@jabber.org/Eve is now online'. Worst case, Alice sees nothing at all (presence notifications in XMPP can be wonky at times). Also imagine that 'Eve' and 'Gajim' look very similar (e.g. substitute 'Eve' for 'Gaijm')
4) Alice sends a message. Since resource has changed (due to Eve's priority being higher than Bob's), OTR session isn't started. If OTR isn't required, this and all further messages from Alice (to bob@jabber.org, which Eve now intercepts) will be sent in plaintext without any warnings.

I have no idea how to fix this *and* keep support for multiple resources.

FWIW, this problem is present in gajim-plugins/gajim-otr
